### ISSUE TITLE
elgg3: elgg-cli cron working correctly?

### ISSUE THREAD

**Issue by iionly** [Elgg/Elgg#12190]
I've tried using the elgg-cli cron command in crontab to let the cronjobs getting triggered without the need to use tokens for the cron urls.

First, I only figured out when testing the command directly on the command line that it does not work without a working cache symlink. Is this missing in the docs or have I just missed this info?

But the Cron page in the admin sections does not show any sign of any cronjobs ever having run. Now I don't know if the crontab might not be triggering them currently (with system-wide crontab you would have to provide the user to be used for execution and a normal user seems not to work and using root seems not a good idea so I'm currently trying to run it as the Apache user wwwrun).

But even so, the Cron page should display at least some intervals as having run, because the manual execution of elgg-cli directly from the command line seems to have resulted in some jobs being run (though it seems to me not all intevals got triggered even with not using the -i parameter).

**Comment by jeabakker** [comment]
> First, I only figured out when testing the command directly on the command line that it does not work without a working cache symlink

Hmmm, wierd we don't have a symlink configured and works perfectly

> But even so, the Cron page should display at least some intervals as having run, because the manual execution of elgg-cli directly from the command line seems to have resulted in some jobs being run (though it seems to me not all intevals got triggered even with not using the -i parameter).

What is the command you are using? 
`elgg-cli cron` should be it. 
That will run all the cron intervals for the current time. Eg. at 14:13 only the minute cron will run, on 14:30 the minute, fiveminute, fifteenminute and halfhour will run

**Comment by iionly** [comment]
For root it works without symlink indeed. I got confused by the error message ("Cache path not writeable") when testing as normal user (without write permissions for data directory). But that's independent of the existence of the cache symlink as you would still have no write access as normal user even with symlink.

When running the elgg-cli command without -q parameter I see the triggered cron intervals on the command line. But I do NOT see anything on the Cron page in the admin section, so I wouldn't know if any jobs have run when triggering is done via cron daemon.

**Comment by iionly** [comment]
I've now tested with the old-fashioned cron urls in crontab and also tested with manually triggering a cron interval via browser. It seems the problem with the Cron page not showing anything is not happening for the elgg-cli approach only but it simply does not show anything even if a cronjob has definitely run.

**Comment by hypeJunction** [comment]
Are you sure there is any output? As @jeabakker mentioned, crons only run once in a given interval, so you can't trigger 5 minute cron at 12:23. 
Try running: ```elgg-cli cron -t "2018-01-01 0:00" -vvv```

**Comment by hypeJunction** [comment]
My table is populated as it should, so there is something wonky with your install. Is that a clean Elgg install? Anything in the logs? Maybe something fatals and crons jobs never complete.

**Comment by iionly** [comment]
Found out what's wrong. The cronjobs did run sort of but...

Running elgg-cli cron as root resulted in the folder /path_to_data_directory/1/1 that hasn't existed before to be created with root as owner. While the logs of the cron interval runs done as root were created, the webserver (wwwrun) couldn't access the folder and therefore showed "Last completed" as "Never". All the following runs done either via crontab entries (as user wwwrun) or triggered in the browser failed then because the existing 1/1 folder wasn't accessible for the Apache user. It showed up by errors in the logs about missing permissions for fopen() and also errors for fwrite() and fclose().

**Comment by iionly** [comment]
Two things that I probably haven't fully understood yet:

- with the cron urls it's possible to use force=1 to trigger a cron interval forcefully. But it seem this does not work when the cron url are secured and tokens are required. Is that true and if yes is that intended?
- I've not yet found out if you can also forcefully trigger a cron interval with elgg-cli. Is it possible and if yes what's the parameter?

**Comment by hypeJunction** [comment]
>with the cron urls it's possible to use force=1 to trigger a cron interval forcefully. But it seem this does not work when the cron url are secured and tokens are required. Is that true and if yes is that intended?

You should be able to pass `force` as query element, which should force cron jobs to run at `* * * * *`. Not sure about the exact logic there. I wonder if that would only force when you run it at exact minute...

>I've not yet found out if you can also forcefully trigger a cron interval with elgg-cli. Is it possible and if yes what's the parameter?

There is no `force` paramter. You can use `-i` (`--interval`) to specify an interval or `-t` (`--time`) to override current time, but using time is discouraged, as it only exists for testing purposes.
http://learn.elgg.org/en/3.0/intro/elgg-cli.html?highlight=cron

**Comment by hypeJunction** [comment]
Try this:

```php
// engine/classes/Elgg/Cron.php:78
if ($force) {
    $cron_interval = self::$intervals['minute'];
    $hours = (int) $time->format('G');
    $minutes = (int) $time->format('i');
    $time->setTime($hours, $minutes, 0);
} else {
    $cron_interval = self::$intervals[$interval];
}
```

**Comment by jdalsem** [comment]
@iionly why did you close this?

**Comment by iionly** [comment]
I don't know what needs to be solved. The original problem I had was caused by calling elgg-cli as root user. This resulted in the folders in the data directory used and not yet existing to be created with root as owner with the consequence that the webserver could access these folders and any cron daemon triggered cronjobs not working.

The force paramater as it is works also. Only that it doesn't work in a way useful for me as it triggers ALL intervals regardless if they are overdue or not. That's not what I was in need for when trying to make use of the elgg-cli code in my crontrigger plugin where I had to find an alternative for the removed run_once function of Elgg.

So the result for me at least is that I most likely will avoid making use of elgg-cli tools as it just doesn't work for me so far without causing problems because I have to use it as root user or it does not work at all. And calling cron via elgg-cli works probably as intended so I closed the issue.
